In [1]:
import pandas as pd
import numpy as np

#### Note :- I have used log base e for the entire assignment.

## Part 1 :-  Tennis dataset.

In [2]:
df = pd.read_csv('play_tennis.csv')

In [3]:
df

day   outlook  temp humidity    wind play
0    D1     Sunny   Hot     High    Weak   No
1    D2     Sunny   Hot     High  Strong   No
2    D3  Overcast   Hot     High    Weak  Yes
3    D4      Rain  Mild     High    Weak  Yes
4    D5      Rain  Cool   Normal    Weak  Yes
5    D6      Rain  Cool   Normal  Strong   No
6    D7  Overcast  Cool   Normal  Strong  Yes
7    D8     Sunny  Mild     High    Weak   No
8    D9     Sunny  Cool   Normal    Weak  Yes
9   D10      Rain  Mild   Normal    Weak  Yes
10  D11     Sunny  Mild   Normal  Strong  Yes
11  D12  Overcast  Mild     High  Strong  Yes
12  D13  Overcast   Hot   Normal    Weak  Yes
13  D14      Rain  Mild     High  Strong   No

### Total entropy of the data set

In [4]:
t_entropy=sum([-x*np.log(x) for x in df['play'].value_counts()/df.shape[0]])
t_entropy

0.6517565611726531

### Entropy of Each columns

In [5]:
final_entropies = {}
for col in df.columns[1:-1]:
    
    ucol=df[col].unique()
    ucol_entropy=[]
    apply=[]
    for x in ucol:
        temp = (df[df[col]==x])
        apply.append(temp.shape[0]/df.shape[0])
        p=temp['play'].value_counts()/temp['play'].value_counts().sum()
        ucol_entropy.append(sum([ -x*np.log(x) for x in p]))
    
    final_entropies[col]=np.dot(apply,ucol_entropy)
final_entropies

{'outlook': 0.48072261929232607,
 'temp': 0.6315010221774208,
 'humidity': 0.5465122114944403,
 'wind': 0.6183974457364384}

### Now, finding the IG (Information Gain) of each columns

In [6]:
ig = {}
for i in final_entropies.keys():
    ig[i]=t_entropy-final_entropies[i]
    print(f'IG of "{i}" column : {ig[i]}')
print(f'\nMaximum IG is {max(ig.values())}')

IG of "outlook" column : 0.17103394188032706
IG of "temp" column : 0.02025553899523236
IG of "humidity" column : 0.10524434967821283
IG of "wind" column : 0.033359115436214726

Maximum IG is 0.17103394188032706


#### We can clearly see that "outlook" column has the max IG. Therefore, the first split will be done on "Outlook" column.

## Part 2 :- Iris dataset.

In [7]:
df=pd.read_csv('iris.csv')
df=df[['petal_length','species']]

In [8]:
df.rename(columns={'petal_length':'pl'},inplace=True)

In [9]:
df

pl    species
0    1.4     setosa
1    1.4     setosa
2    1.3     setosa
3    1.5     setosa
4    1.4     setosa
..   ...        ...
145  5.2  virginica
146  5.0  virginica
147  5.2  virginica
148  5.4  virginica
149  5.1  virginica

[150 rows x 2 columns]

In [10]:
def find_split(df,scol,tcol):
        
    # Total entropy of parent dataset.
    t_entropy=sum([-i*np.log(i) for i in df[tcol].value_counts()/df.shape[0]])
    
    ig = {}
    
    # Sorting the spliting column.
    df.sort_values(by=scol,inplace=True)
    
    for x in df[scol][:-1]:
        
        # Spliting the dataset for each value in spliting column.
        df1=df[df[scol]<=x]
        df2=df[df[scol]>x]

        # Calculating the entripies of each splited dataset.
        e1=sum([-i*np.log(i) for i in df1[tcol].value_counts()/df1[tcol].shape[0]])
        e2=sum([-i*np.log(i) for i in df2[tcol].value_counts()/df2[tcol].shape[0]])

        # Calculating the IG for splitting value
        ig[x]=t_entropy-((df1.shape[0]/df.shape[0])*e1+(df2.shape[0]/df.shape[0])*e2)
    
    return max(ig,key=ig.get)

In [11]:
find_split(df,'pl','species')

1.9

### Therefore, the spliting will be done at petal_length > 1.9